### PR TITLE
[WasmFS] Implement O_TRUNC support in open

### PR DESCRIPTION
**Pull request by tlively**

### CODE REVIEW

File: system/lib/wasmfs/syscalls.cpp [hunk 1]
```diff
@@ -465,6 +465,12 @@ static __wasi_fd_t doOpen(path::ParsedParent parsed,
     return -EEXIST;
   }
 
+  // If O_TRUNC, truncate the file if possible.
+  if (flags & O_TRUNC && child->is<DataFile>() &&
+      fileMode & WASMFS_PERM_WRITE) {
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:470]: Do we not need `()` around the `&` operation and its arguments? At least in JS I believe we do. For clarity I'd add it here.

Also, if we lack write permissions should we not error?

File: tests/unistd/truncate.c [hunk 2]
```diff
@@ -79,15 +79,24 @@ int main() {
...
+  fstat(f, &s);
   printf("st_size: %lld\n", s.st_size);
   memset(&s, 0, sizeof s);
   errno = 0;
   printf("\n");
 
   printf("truncate(readonly, 2): %d\n", truncate("toread", 2));
   printf("errno: %s\n", strerror(errno));
-  stat("toread", &s);
+  fstat(f2, &s);
```
Comment (kripken) [inline tests/unistd/truncate.c:99]: Why not leave one check for `stat` for more coverage?

File: tests/unistd/truncate.c [hunk 3]
```diff
@@ -99,6 +108,15 @@ int main() {
   printf("st_size: %lld\n", s.st_size);
   memset(&s, 0, sizeof s);
   errno = 0;
+  printf("\n");
+
+  printf("open(readonly, O_TRUNC)\n");
+  open("toread", O_RDONLY | O_TRUNC);
```
Comment (kripken) [inline tests/unistd/truncate.c:114]: Is this not the same as above? Both open in read-only mode + trunc. Should the former be read-write or write-only?

File: system/lib/wasmfs/syscalls.cpp (hunk 1 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:470]: I can add parens, although my compiler didn’t mind me leaving them out. O_TRUNC is ignored if the file is not writable.

File: tests/unistd/truncate.c (hunk 3 above)
Comment (tlively) [reply tests/unistd/truncate.c:114]: It is the same and that is intentional. The behavior should depend on the permissions of the file, not the mode we are opening it in.

File: tests/unistd/truncate.c (hunk 2 above)
Comment (kripken) [reply tests/unistd/truncate.c:99]: sgtm, but please confirm `stat` is actually covered. I don't remember offhand.

File: tests/unistd/truncate.c (hunk 3 above)
Comment (kripken) [reply tests/unistd/truncate.c:114]: I see, sounds good. Maybe add a comment here for that?

**Review by kripken — APPROVED** [review]
lgtm % comments

File: tests/unistd/truncate.c (hunk 2 above)
Comment (tlively) [reply tests/unistd/truncate.c:99]: Confirmed! We even have a whole test directory tests/stat.

**Comment by tlively** [comment]
Changing the behavior to return EACCES if the file is not writable to match the current FS and Linux (the docs say it is unspecified behavior).

**Review by kripken — APPROVED** [review]

**Comment by tlively** [comment]
Merging despite unrelated-looking browser test failures

**Pull request merged**